### ISSUE TITLE
ddp got stuck(deadlock)

### ISSUE THREAD

**Issue by kwonjunn01** [pytorch/pytorch#85334]
### 🐛 Describe the bug

Hi!

I'm running mmsegmentation code.
However, when running the program, the program stops randomly even in manual_seed state.
As a result of searching related materials, it seems to be a ddp related issue.
When the program is tested in a single gpu state, it runs without any problem, but in a multi-gpu situation, there is a problem in execution.
In this case, the computer must be rebooted by removing the power.
AMD CPU + A6000 combination, iommu is off.
NCCL_P2P_DISABLE also tried all cases.
Also, when learning on each gpu with single gpu training, it falls into a deadlock state.
### Versions

sys.platform: linux
Python: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:18) [GCC 10.3.0]
CUDA available: True
GPU 0,1,2: NVIDIA RTX A6000
CUDA_HOME: /usr/local/cuda
NVCC: Build cuda_11.4.r11.4/compiler.30033411_0
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.10.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.11.2
OpenCV: 4.6.0
MMCV: 1.3.17
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.3
MMSegmentation: 0.19.0+3b72447

Below is the nccl debug info.
-----------------------------------------------------------------------------------------------------------------------------
a8:2963:2963 [0] NCCL INFO Bootstrap : Using enp37s0f1:165.132.57.160<0>
a8:2963:2963 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation

a8:2963:2963 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
a8:2963:2963 [0] NCCL INFO NET/Socket : Using [0]enp37s0f1:165.132.57.160<0>
a8:2963:2963 [0] NCCL INFO Using network Socket
NCCL version 2.10.3+cuda11.3
a8:2964:2964 [1] NCCL INFO Bootstrap : Using enp37s0f1:165.132.57.160<0>
a8:2964:2964 [1] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation

a8:2964:2964 [1] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
a8:2964:2964 [1] NCCL INFO NET/Socket : Using [0]enp37s0f1:165.132.57.160<0>
a8:2964:2964 [1] NCCL INFO Using network Socket
a8:2964:3089 [1] NCCL INFO Trees [0] -1/-1/-1->1->0 [1] 0/-1/-1->1->-1 [2] -1/-1/-1->1->0 [3] 0/-1/-1->1->-1
a8:2963:3088 [0] NCCL INFO Channel 00/04 :    0   1
a8:2963:3088 [0] NCCL INFO Channel 01/04 :    0   1
a8:2963:3088 [0] NCCL INFO Channel 02/04 :    0   1
a8:2963:3088 [0] NCCL INFO Channel 03/04 :    0   1
a8:2963:3088 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1 [1] -1/-1/-1->0->1 [2] 1/-1/-1->0->-1 [3] -1/-1/-1->0->1
a8:2964:3089 [1] NCCL INFO Channel 00 : 1[41000] -> 0[2e000] via P2P/IPC
a8:2964:3089 [1] NCCL INFO Channel 01 : 1[41000] -> 0[2e000] via P2P/IPC
a8:2963:3088 [0] NCCL INFO Channel 00 : 0[2e000] -> 1[41000] via P2P/IPC
a8:2964:3089 [1] NCCL INFO Channel 02 : 1[41000] -> 0[2e000] via P2P/IPC
a8:2963:3088 [0] NCCL INFO Channel 01 : 0[2e000] -> 1[41000] via P2P/IPC
a8:2964:3089 [1] NCCL INFO Channel 03 : 1[41000] -> 0[2e000] via P2P/IPC
a8:2963:3088 [0] NCCL INFO Channel 02 : 0[2e000] -> 1[41000] via P2P/IPC
a8:2963:3088 [0] NCCL INFO Channel 03 : 0[2e000] -> 1[41000] via P2P/IPC
a8:2964:3089 [1] NCCL INFO Connected all rings
a8:2964:3089 [1] NCCL INFO Connected all trees
a8:2964:3089 [1] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/512
a8:2964:3089 [1] NCCL INFO 4 coll channels, 4 p2p channels, 2 p2p channels per peer
a8:2963:3088 [0] NCCL INFO Connected all rings
a8:2963:3088 [0] NCCL INFO Connected all trees
a8:2963:3088 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/512
a8:2963:3088 [0] NCCL INFO 4 coll channels, 4 p2p channels, 2 p2p channels per peer
a8:2964:3089 [1] NCCL INFO comm 0x7fbf08002f70 rank 1 nranks 2 cudaDev 1 busId 41000 - Init COMPLETE
a8:2963:3088 [0] NCCL INFO comm 0x7fd9a4002f70 rank 0 nranks 2 cudaDev 0 busId 2e000 - Init COMPLETE
a8:2963:2963 [0] NCCL INFO Launch mode Parallel

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by kwen2501** [comment]
Hi, I am glad that you looked into IOMMU.

In your description:
> NCCL_P2P_DISABLE also tried all cases.

Just for our better understanding: does setting `NCCL_P2P_DISABLE=1` work or not?

If `NCCL_P2P_DISABLE=1` works but `NCCL_P2P_DISABLE=0` doesn't, then this could still be a VT-d or IOMMU issue. In that case, I recommend:
- cross check with [this section](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html#pci-access-control-services-acs) in NCCL's troubleshooting doc
- download and build NCCL test from [here](https://github.com/NVIDIA/nccl-tests) and see if it works.

**Comment by kwonjunn01** [comment]
Hi thank you for your reply @kwen2501 

Under setting  '''NCCL_P2P_DISABLE=1'''', progress does not work. 
And remove  '''NCCL_P2P_DISABLE=1(# NCCL_P2P_DISABLE=1 )'''', in this situation, progress doesn't work too. 

Although I format the server and reinstall all requirements for my project,  This also doesn't work.   



Inspection results referring to the NCCL documentation, it says that acs is off.

ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-
                ACSCtl: SrcValid- TransBlk- ReqRedir- CmpltRedir- UpstreamFwd- EgressCtrl- DirectTrans-

it is my nccl-test results

# nThread 1 nGpus 3 minBytes 8 maxBytes 134217728 step: 2(factor) warmup iters: 5 iters: 20 agg iters: 1 validation: 1 graph: 0
#
# Using devices
#   Rank  0 Pid  28223 on      rader device  0 [0x2e] NVIDIA RTX A6000
#   Rank  1 Pid  28223 on      rader device  1 [0x41] NVIDIA RTX A6000
#   Rank  2 Pid  28223 on      rader device  2 [0x61] NVIDIA RTX A6000
#
#                                                              out-of-place                       in-place          
#       size         count      type   redop    root     time   algbw   busbw #wrong     time   algbw   busbw #wrong
#        (B)    (elements)                               (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
           8             2     float     sum      -1     8.64    0.00    0.00      0     8.92    0.00    0.00      0
          16             4     float     sum      -1     8.66    0.00    0.00      0     8.70    0.00    0.00      0
          32             8     float     sum      -1     8.61    0.00    0.00      0     8.60    0.00    0.00      0
          64            16     float     sum      -1     8.67    0.01    0.01      0     8.64    0.01    0.01      0
         128            32     float     sum      -1     8.71    0.01    0.02      0     8.89    0.01    0.02      0
         256            64     float     sum      -1     8.67    0.03    0.04      0     8.70    0.03    0.04      0
         512           128     float     sum      -1     8.70    0.06    0.08      0     8.78    0.06    0.08      0
        1024           256     float     sum      -1     8.72    0.12    0.16      0     8.69    0.12    0.16      0
        2048           512     float     sum      -1     8.94    0.23    0.31      0     8.68    0.24    0.31      0
        4096          1024     float     sum      -1     9.19    0.45    0.59      0     8.80    0.47    0.62      0
        8192          2048     float     sum      -1     9.68    0.85    1.13      0     9.48    0.86    1.15      0
       16384          4096     float     sum      -1    10.61    1.54    2.06      0    10.81    1.52    2.02      0
       32768          8192     float     sum      -1    16.12    2.03    2.71      0    16.01    2.05    2.73      0
       65536         16384     float     sum      -1    29.29    2.24    2.98      0    28.84    2.27    3.03      0
      131072         32768     float     sum      -1    43.84    2.99    3.99      0    43.07    3.04    4.06      0
      262144         65536     float     sum      -1    67.41    3.89    5.19      0    68.37    3.83    5.11      0
      524288        131072     float     sum      -1    95.83    5.47    7.29      0    99.34    5.28    7.04      0
     1048576        262144     float     sum      -1    129.1    8.12   10.83      0    134.9    7.77   10.37      0
     2097152        524288     float     sum      -1    185.2   11.33   15.10      0    189.5   11.07   14.76      0
     4194304       1048576     float     sum      -1    324.2   12.94   17.25      0    308.5   13.60   18.13      0
     8388608       2097152     float     sum      -1    577.8   14.52   19.36      0    573.6   14.62   19.50      0
    16777216       4194304     float     sum      -1   1121.8   14.96   19.94      0   1119.7   14.98   19.98      0
    33554432       8388608     float     sum      -1   2231.9   15.03   20.04      0   2233.3   15.02   20.03      0
    67108864      16777216     float     sum      -1   4451.7   15.08   20.10      0   4441.2   15.11   20.15      0
   134217728      33554432     float     sum      -1   8882.6   15.11   20.15      0   8891.5   15.10   20.13      0
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 6.77498 
#
There doesn't seem to be any other problems in my opinion.
And it is my topology matrix
        GPU0    GPU1    GPU2    CPU Affinity    NUMA Affinity
GPU0     X      SYS     SYS     0-31            N/A
GPU1    SYS      X      SYS     0-31            N/A
GPU2    SYS     SYS      X      0-31            N/A

I haven't solved the problem yet.
The odd thing is that on other servers with the same configuration, the program runs without problems.
May I know what's the problem?

**Comment by kwonjunn01** [comment]
I solved the problem!!


It was a gpu power cable issue!!

**Comment by kwen2501** [comment]
Glad that the issue is resolved. I am closing it if you don't mind.
